### PR TITLE
dialects: Make passes a type

### PR DESCRIPTION
**Pull request by PapyChacal**
This PR offers to: 

- Switch passes from simple fixed-signature functions to types, to simply add parameters later-on, and probably desired at some point for analysis?
- Make a pass' name part of itself rather than part of xdsl-opt.
- ~~Make passes act on generic Operation rather than ModuleOp~~

I guess all those could be split, if some idea meet more resistance than others! My main motivation is options on passes, so making them types or complexifying the signature if we prefer that.

Next-step goals motivating me to do so:

- Options on passes
- Pass pipeline builder ? similar to MLIR's [textual pass pipeline specification](https://mlir.llvm.org/docs/PassManagement/#textual-pass-pipeline-specification)

### CODE REVIEW

File: xdsl/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from abc import abstractmethod
+from xdsl.ir import MLContext, Operation
+
+
+class OperationPass:
```
Comment (webmiche) [inline xdsl/pipeline.py:5]: What is the guarantee of an OperationPass? Is it executed on each Operation? AFAIU, the classic LLVM passes are Module/Function/Loop passes, all of which are run on individual Modules/Functions/Loops, making them inherently parallel. Now, we don't actually do parallelism at this point in Python land, but we do require that each pass is run on valid IR. This is something I don't quite get from this pass. What is it supposed to be applied on?

File: xdsl/xdsl_opt_main.py [hunk 2]
```diff
@@ -222,19 +223,20 @@ def parse_mlir(io: IO[str]):
...
         Add other/additional passes by overloading this function.
         """
-        self.available_passes['lower-mpi'] = lower_mpi
-        self.available_passes[
-            'convert-stencil-to-ll-mlir'] = ConvertStencilToLLMLIR
-        self.available_passes['convert-stencil-to-gpu'] = ConvertStencilToGPU
-        self.available_passes[
-            'stencil-shape-inference'] = StencilShapeInference
-        self.available_passes['frontend-desymrefy'] = Desymrefy
+        self.register_pass(LowerMPIPass)
```
Comment (webmiche) [inline xdsl/xdsl_opt_main.py:235]: I personally preferred to old way. This gave you the information of how to call a given class in more straight-forward, central manner, IMO.

File: xdsl/xdsl_opt_main.py [hunk 3]
```diff
@@ -2,11 +2,12 @@
 import sys
 import os
 from io import StringIO
-from xdsl.frontend.passes.desymref import Desymrefy
+from xdsl.frontend.passes.desymref import DesymrefyPass
```
Comment (webmiche) [inline xdsl/xdsl_opt_main.py:5]: Can you order these a bit? Maybe group all passes?

File: xdsl/transforms/lower_mpi.py [hunk 4]
```diff
@@ -742,7 +743,11 @@ def walker(op: Operation):
...
 
 
-def lower_mpi(ctx: MLContext, module: builtin.ModuleOp):
+@dataclass
+class LowerMPIPass(OperationPass):
+
+    name = 'lower-mpi'
+
     # TODO: how to get the lib info in here?
     lib_info = MpiLibraryInfo()
```
Comment (webmiche) [inline xdsl/transforms/lower_mpi.py:752]: Not sure about this, but don't we run into the classic Python shared argument problem here? If you instantiate two of these and mutate this on one, is it visible from the other one?

**Review by webmiche — COMMENTED** [review]
Thanks for the PR! I think there is a lot of things we should discuss about our rewriting/pass infrastructure. This already brings a lot of this up, cool :)

I am a bit torn on the Pass as a class idea. I personally think of passes more as function than as classes and I don't see a big advantage of changing it other than having access to fields. But, IMO, if a pass needs to preserve state over multiple executions, there are other ways to do this...

**Comment by webmiche** [comment]
Maybe we can make passes have an option_list argument to allow for options?

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 5]
```diff
@@ -7,12 +7,13 @@
                                    op_type_rewrite_pattern)
 from xdsl.ir import BlockArgument, MLContext, Operation
 from xdsl.irdl import Attribute
-from xdsl.dialects.builtin import FunctionType, ModuleOp
+from xdsl.dialects.builtin import FunctionType
 from xdsl.dialects.func import FuncOp
 from xdsl.dialects.memref import MemRefType
 from xdsl.dialects import memref, arith, scf, builtin, gpu
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:13]: Are you supposed to use memref.MemrefType so as to avoid this input?

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py
Comment (georgebisbas) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:1]: Overall I feel docstrings are missing

File: xdsl/transforms/lower_mpi.py (hunk 4 above)
Comment (PapyChacal) [reply xdsl/transforms/lower_mpi.py:752]: It is, yes. Not sure if that's something that we should be concerned about though : it is then registered under the initial name.

Comment (georgebisbas) [inline xdsl/transforms/lower_mpi.py:752]: what info exactly do you need?

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:235]: I get that. Note that it is displayed in `xdsl-opt --help`

File: xdsl/xdsl_opt_main.py [hunk 6]
```diff
@@ -2,11 +2,12 @@
 import sys
 import os
 from io import StringIO
```
Comment (georgebisbas) [inline xdsl/xdsl_opt_main.py:4]: emptyline

**Review by georgebisbas — CHANGES_REQUESTED** [review]
Overall, I feel it is going the right direction, some things here and there. Thanks

File: xdsl/pipeline.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/pipeline.py:5]: A lot of legit concerns. My current answer would be "whatever guarantees our current anonymous functions have".
So, your main concern on this point is acting on a generic Operation rather than on the ModuleOp level? I can switch back to ModuleOp for the moment. I was simply looking forward to nested passes.

File: xdsl/transforms/lower_mpi.py (hunk 4 above)
Comment (PapyChacal) [reply xdsl/transforms/lower_mpi.py:752]: I don't, this comment was there and is about the MPI lowering itself, not the refactoring of passes in general

File: xdsl/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+    name: str
+
+    @abstractmethod
+    def apply(self, ctx: MLContext, op: Operation) -> None:
+        pass
```
Comment (superlopuh) [inline xdsl/pipeline.py:11]:
```suggestion
        ...
```

**Review by superlopuh — APPROVED** [review]
Love this, structures over functions always :)

File: xdsl/transforms/lower_mpi.py (hunk 4 above)
Comment (webmiche) [reply xdsl/transforms/lower_mpi.py:752]: I was talking about the lib_info, which seems more dangerous to me.

File: xdsl/xdsl_opt_main.py [hunk 8]
```diff
@@ -60,7 +61,7 @@ class xDSLOptMain:
     stream.
     """
 
-    pipeline: List[tuple[str, Callable[[ModuleOp], None]]]
+    pipeline: List[OperationPass]
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:64]: Exactly my secret plan :yum:
Waiting for eventually more feedback, and maybe this needs to be taken to Zulip if it starts a huge thing

**Comment by webmiche** [comment]
> > Maybe we can make passes have an option_list argument to allow for options?
> 
> We could, sure, and I'd be happy to use that as first steps. I just think types are clearer and more readable through the codebase? At some point, we probably (I at least do) want to compose passes and nest them (e.g., apply this pass only on that operation type nested in this other type). going for functions only means a lot of involved functional constructions. I'm not afraid of those but I'm not sure they would be clearer, especially for debugging.

I don't think you can nest/compose in a pass-based infrastructure. I am down to explore other infrastructures as well, in particular I think that passes are too coarse-grained in LLVM and don't give the necessary freedom to really do compilation. But I guess we are running into PR creep again.

I am fine with making stuff classes, I jsut think we need to get the right semantics and guarantees down for the different passes, which is something that I am missing atm.

File: xdsl/transforms/lower_mpi.py (hunk 4 above)
Comment (superlopuh) [inline xdsl/transforms/lower_mpi.py:752]:
```suggestion
    lib_info: MpiLibraryInfo = field(default_factory=MpiLibraryInfo)
```

Comment (PapyChacal) [reply xdsl/transforms/lower_mpi.py:752]: Oh! Maybe this should just be in the function then, yes.
@AntonLydike, do you have an input on that?

File: xdsl/pipeline.py (hunk 1 above)
Comment (webmiche) [reply xdsl/pipeline.py:5]: Yes, I think making this a `ModulePass` is the way to go for the moment :)

We can discuuss nesting/composing in next PRs/discussions.

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:235]: Exactly, I looked into printing more complete help there too, just gave up on output formatting details because all that is already a sidetrack for me.
But then, we could have something like `xdsl-opt --help passes` displaying extensive information from docstring or whatever we think makes sense

File: xdsl/pipeline.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/pipeline.py:5]: I'm not sure what you mean by guarantees, do you mean what is expected by implementations of this class? Like leaving the IR in a verifiable state?

**Comment by PapyChacal** [comment]
> > > Maybe we can make passes have an option_list argument to allow for options?
> > 
> > 
> > We could, sure, and I'd be happy to use that as first steps. I just think types are clearer and more readable through the codebase? At some point, we probably (I at least do) want to compose passes and nest them (e.g., apply this pass only on that operation type nested in this other type). going for functions only means a lot of involved functional constructions. I'm not afraid of those but I'm not sure they would be clearer, especially for debugging.
> 
> I don't think you can nest/compose in a pass-based infrastructure. I am down to explore other infrastructures as well, in particular I think that passes are too coarse-grained in LLVM and don't give the necessary freedom to really do compilation. But I guess we are running into PR creep again.
> 
> I am fine with making stuff classes, I jsut think we need to get the right semantics and guarantees down for the different passes, which is something that I am missing atm.

My specific vocabulary is probably still a bit off, but I know we can nest/compose in a pass-based infrastructure, at least what I call that way :sweat_smile: I'm just aiming at something analog to the textual pass pipeline of MLIR, which is finer grained that what we have already, given that you can restrict some passes to some nested ops.
Now, I completely agree, we need to go further than that at some point. But starting from here sounds like a good first step to me.

File: xdsl/pipeline.py (hunk 1 above)
Comment (webmiche) [reply xdsl/pipeline.py:5]: Yes, things like "consume valid IR and produces valid IR".

**Comment by webmiche** [comment]
AFAIK, nesting/composing passes is very dangerous because IR might be invalid during the execution of a Pass. I guess if you make sure you have valid IR, you might be able to do something though.

Sure, let's aim at that first and go from there.

WHat do you think about an ABC `Pass` and then `ModulePass` is just an implementation of it?

File: xdsl/transforms/lower_mpi.py (hunk 4 above)
Comment (webmiche) [reply xdsl/transforms/lower_mpi.py:752]: This will fix my issue for the MpiLibraryInfo, but the walkers are still an issue, AFAIU.

**Comment by PapyChacal** [comment]
Completely okay with your propsal! So I'll bar out "going from ModuleOp to Operation" from this PR and adress other comments soon, I'm busy on other tasks right now. Switching to draft to make it clear that this PR is on standby mode for now :)

**Comment by webmiche** [comment]
Sure, thanks :+1:

File: xdsl/xdsl_opt_main.py (hunk 3 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py]: I could group the import lines, but to do a single line, we would have to reorganize them in the filesystem AFAIU.

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:1]: From this lowerings for sure, I'll put some effort on that direction too now, but not in this PR.

File: xdsl/xdsl_opt_main.py (hunk 6 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:4]: What did you mean?

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:235]: @webmiche, are you okay with this change or are you strongly for keeping the names here?

File: xdsl/xdsl_opt_main.py [hunk 9]
```diff
@@ -225,21 +226,21 @@ def parse_mlir(io: IO[str]):
...
-            'stencil-shape-inference'] = StencilShapeInference
-        self.available_passes[
-            'stencil-to-local-2d-horizontal'] = global_stencil_to_local_stencil_2d_horizontal
-        self.available_passes['frontend-desymrefy'] = Desymrefy
+        self.register_pass(LowerMPIPass)
+        self.register_pass(ConvertStencilToLLMLIRPass)
+        self.register_pass(ConvertStencilToGPUPass)
+        self.register_pass(StencilShapeInferencePass)
+        self.register_pass(GlobalStencilToLocalStencil2DHorizontal)
+        self.register_pass(DesymrefyPass)
```
Comment (AntonLydike) [inline xdsl/xdsl_opt_main.py:243]: This looks a lot nicer to me, love it :heart_eyes:

**Review by AntonLydike — APPROVED** [review]
LGTM, this is exciting! Next step is pass options? :eyes:

**Review by math-fehr — APPROVED** [review]
👍

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 10]
```diff
@@ -32,21 +32,22 @@
 
 def GetMemRefFromField(
     input_type: FieldType[_TypeElement] | TempType[_TypeElement]
-) -> MemRefType[_TypeElement]:
+) -> memref.MemRefType[_TypeElement]:
```
Comment (webmiche) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:35]: These changes feel unrelated, right? This is a Stencil refactor, AFAIU.

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (webmiche) [reply xdsl/xdsl_opt_main.py:238]: I am fine with it, as long as we add the necessary --help support for it in the near future.

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 11]
```diff
@@ -435,39 +436,48 @@ def StencilConversion(return_targets: dict[ReturnOp, CastOp | memref.Cast],
     ])
 
 
-# TODO: We probably want to factor that in another file
-def StencilShapeInference(ctx: MLContext, module: ModuleOp):
+class StencilShapeInferencePass(ModulePass):
+
+    name = 'stencil-shape-inference'
+
+    inference_walker = PatternRewriteWalker(ShapeInference,
```
Comment (webmiche) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:443]: I generally get a weird feeling when defining fields/functions with default objects in Python, so I would generally prefer to move this into the apply function.

In a later PR, we might also want to integrate this more deeply with our current rewriting infrastructure, s.t. you could for example override the `walker` field, or a `Pattern` list and then you don't have to even give an apply method. Let's leave this for later though :)

File: xdsl/pipeline.py
Comment (webmiche) [inline xdsl/pipeline.py:1]: I personally dislike the word pipeline for this as it is extremely overused in computer science, IMO. I know LLVM uses it as well, but wouldn't just `pass.py` already describe what is in here?

Comment (webmiche) [reply xdsl/pipeline.py:1]: Maybe a clearer comment: When talking to people outside of the compilation field, the usage of the word "pipeline" can lead to confusion as it is usually tied to the concept of this multi-stage thing, where each stage can be working on a different operation at any given time, so more in the processor pipelining way. In compilation, we seem to usually use it in a  "this comes after that" sense.

In my Master's thesis, I tried to replace the "compilation pipeline" with the "compilation stack" (because it goes from top down through abstraction levels) or "pass hierarchy"/"pass structure"/"pass schedule". IMO, the last of these is really what this is all about and describes it best.

**Review by webmiche — CHANGES_REQUESTED** [review]
After some more thinking, I feel like we can do this change as it is mainly a change in how xDSL opt works (making it clearer) and not really something fundamental about how you can rewrite IR in the project.

That said, I think there are a couple of changes still needed though.

File: xdsl/xdsl_opt_main.py (hunk 6 above)
Comment (georgebisbas) [reply xdsl/xdsl_opt_main.py:4]: To add an empty line between lined 3 and 4

**Review by georgebisbas — APPROVED** [review]

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 11 above)
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:443]: Sure, leaving things out of apply was more a style thing to me, I can leave everything in for now.

+10 for the deeper integration ☺️

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 10 above)
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:35]: True, will revert those

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:238]: Note that xdsl-opt --help already prints all available passes' names.
It would be lovely to print more information somewhere there, such as the docstring if present? I will see if I can do something on that side.

File: xdsl/pipeline.py
Comment (PapyChacal) [reply xdsl/pipeline.py:1]: I can rename the file to pass.py, sure.
Completely agree on the confusing aspect of those terms. I just mimiced MLIR here, but I value thinking the vocabulary again at some point. This will be a tradeoff again for xDSL, in "feeling home to MLIR people" vs try to improve this aspect from our side.
My personal opinion is we should take the liberty to change those things in xDSL if we feel it makes sense, especially for similar concepts like those, where it's easy to just document "It is similar to what MLIR calls a pipeline". I don't think it is technically orthogonal to interoperability too, but this is where changing things from our end could get even more confusing..
I will just rename the file for this PR, and I think this might be a nice topic to discuss more broadly

Comment (PapyChacal) [reply xdsl/pipeline.py:1]: Oh, right, pass is a python keyword, so `from xdsl.pass import [...]` is invalid syntax :sweat_smile: 
Renaming to passes.py..

Comment (webmiche) [reply xdsl/pipeline.py:1]: I mean it contains the ModulePass, right? I think passes is fine for this :)

File: xdsl/xdsl_opt_main.py [hunk 12]
```diff
@@ -61,7 +63,7 @@ class xDSLOptMain:
     stream.
     """
 
-    pipeline: List[tuple[str, Callable[[ModuleOp], None]]]
+    pipeline: List[ModulePass]
```
Comment (webmiche) [inline xdsl/xdsl_opt_main.py:66]: Oh no, the pipeline already infected xDSL :fearful:

File: xdsl/pipeline.py
Comment (webmiche) [reply xdsl/pipeline.py:1]: Maybe open an issue or a discussion though?

File: xdsl/xdsl_opt_main.py (hunk 12 above)
Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:66]: Ahahah good catch! It's just a list of passes for now, so I don't mind renaming it `passes`.

Comment (webmiche) [reply xdsl/xdsl_opt_main.py:66]: I don't mind, it was there before, so it can stay (for the time being)...

Comment (PapyChacal) [reply xdsl/xdsl_opt_main.py:66]: Oh, forgot it was already there and named in functions and everything.. I'd better leave it as is for this PR then!
